One of the most common things we do in Angular services is making HTTP calls. And it is very important to write unit tests for the code that makes api requests using the Angular's HTTPClient service.

In our project, we have an ApiService that has two methods, one using the get() method and the other using the post() method of the HttpClient service.

So, let's create the spec.ts file for this Service.

Just like in case of the previous spec.ts file, we will have to first configure a Testing module.

    describe('APIService', () => { 

        let apiService: ApiService;
        beforeEach(() => {
            TestBed.configureTestingModule({
                providers: [ApiService]
            });

            apiService = TestBed.inject(ApiService);
        });

        it('creates a service', () => {
            expect(apiService).toBeTruthy();
        });

    });

But, when we run "npm run test" we see that the console throws an error. This is something we have seen before - 

    No provider for _HttpClient!

Since we are using another service inside the APIService, we have to also ensure that it is provided in the Testing Module. We can add HttpClient in the providers array of the Testing Module but it does not make sense because we just want to test the code and not make actual API calls in our testing environment. If we make actual API calls, the response time may be too long but we want our tests to be quick. And on top of all this, we want to test in isolation. So, API calls are something we do not want to do in testing environment.

So, what can we do? Well, there is something called "HttpClientTestingModule" that Angular provides. This makes testing scenarios that include HttpClient very easy. Otherwise, the testing code can get complicated with mocks and spies.

We have to import this "HttpClientTestingModule" in our Testing module so we do that using the "imports" property.

    TestBed.configureTestingModule({
        imports: [HttpClientTestingModule],
        providers: [ApiService]
    });

Note that the "HttpClientTestingModule" is imported like this - 

    import {HttpClientTestingModule} from "@angular/common/http/testing"

Now, when we save the file, we see that all the tests pass which means the ApiService instance has been created successfully.

So, we can start testing the methods now.

But, there is still one thing to do, something we did not do before. 

We also have to use something called "HttpTestingController" which makes it easy to mock requests. 

And we can get hold of its instance using TestBed.inject - 

    let httpTestingController: HttpTestingController;
    beforeEach(() => {
            TestBed.configureTestingModule({
                imports: [HttpClientTestingModule],
                providers: [ApiService]
            });

            apiService = TestBed.inject(ApiService);
            httpTestingController = TestBed.inject(HttpTestingController);
        });

If you hover over the "HttpTestingController", it will say -

    Controller to be injected into tests, that allows for mocking and flushing of requests.

Now, after each test, wealso have to verify that there are no unmatched requests that are outstanding. That is everything is resetted after a test finishes. And we can do that using an "afterEach()" function.

    afterEach(() => {
        httpTestingController.verify();
    });

This "verify()" method will verify that no unmatched requests are outstanding. If any requests are outstanding, it will fail with an error message indicating which requests were not handled.

So, let's write our first test for the getTags() method - 

    it('should return a list of tags', () => {
        apiService.getTags().subscribe(response => {
            expect(response).toEqual([]);
        });
    });

Since getTags() returns an Observable, we can subscribe to it and add an expect inside it. Now, let's save and see what happens.

We see that the console throws an error which says - 

    "Expected no open requests, found 1: GET http://localhost:3004/tags

This is an error thrown by the "verify()" method of the HttpTestingController. This is thrown because when we call the getTags() method of the APIService, it is making an actual API call to the above url. And we do not really want to make an actual API call in our tests. And this is how the "verify()" method can help to identify such places in the testing file.

So, somehow, we want to mock this get request that we are making inside getTags() method.

And we can do that using the "HttpTestingController". 

There is a method called "expectOne" on this that we can use and this method accepts the URL as an input. This method will then return the mock of the request to that URL.

If no such request has been made, or more than one such request has been made, it will fail with an error message including the given request description, if any.

    it('should return a list of tags', () => {

        apiService.getTags().subscribe(response => {
            expect(response).toEqual([]);
        });

        const req = httpTestingController.expectOne('http://localhost:3004/tags');

    });

Note that we have used this "expectOne" function after the subscribe, not before it. That's because we first want to make a call to getTags() so that a request to the specific url is made and then the HttpTestingController then returns a mock of this request. So, it will only give us a mock of a request, if a request to the specific URL has been made. It means, we had added any other url that is not called in the get() call, the test will still fail.

Now, we also want to mock the response. We can do that using this "req" variable that we have - 

    const req = httpTestingController.expectOne('http://localhost:3004/tags');

    req.flush([{id: '1', name: 'Foo'}]);

So now, when the call to this URL is made, a mock response is provided in the test file that returns an array with one Tag in it.

But there is one issue.

We have this inside subscribe - 

    apiService.getTags().subscribe(response => {     
        expect(response).toEqual([]);
    });

But since we specifically mentioned that the response will be an array with one object, this test should fail right? But it does not fail. It passes.

Let's try to console.log the response to see what are we getting as respones - 

    apiService.getTags().subscribe(response => {
        console.log("Response is", response)
        expect(response).toEqual([]);
    });

We will see that as a response, we are indeed getting an array with one object. Then, the test should fail because "expect" does not match.

But why is it passing?

This is one important thing to keep in mind when using "subscribe()" inside test files. An "expect" inside a subscribe() method will always pass, doesn't matter what is the expected result.

Even if we write - 

    apiService.getTags().subscribe(response => {
        expect(response).toEqual("asfhajkahsjkhd");
    });

This will also pass.

Ofcourse this is definitely not what we want.

And the workaround is to create a variable outside the subscribe and set its value as the response inside in the subscribe.

So, our test now becomes - 


    it('should return a list of tags', () => {

        let tags: TagInterface[] | undefined;
        apiService.getTags().subscribe(response => {
            tags = response;
        });

        const req = httpTestingController.expectOne('http://localhost:3004/tags');

        req.flush([{id: '1', name: 'Foo'}]);

        expect(tags).toEqual([]);
    });

Note how now, we are adding an "expect()" at the end where we check if "tags" is an empty list. Now, this test should fail because we are not getting an empty list as response but rather a list with an object. And indeed it fails.

So, we can now update the test so that it passes - 

    expect(tags).toEqual([{id: '1', name: 'Foo'}]);

And now you can try to write a test for the "createTag()" method which will be quite similar as the test for the "getTags()" method.

